File: section2/116.填充每个节点的下一个右侧节点指针.py
#
# @lc app=leetcode.cn id=116 lang=python3
#
# [116] 填充每个节点的下一个右侧节点指针
#

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""


class Solution:
    def connect_recursive(self, root, next_root):
        if root:
            root.next = next_root
            self.connect_recursive(root.left, root.right)
            self.connect_recursive(root.right, next_root.left if next_root else None)

    def connect(self, root: 'Node') -> 'Node':
        self.connect_recursive(root, None)

        return root

# @lc code=end

File: section2/section/116.填充每个节点的下一个右侧节点指针.go
package section

import "fmt"

/*
 * @lc app=leetcode.cn id=116 lang=golang
 *
 * [116] 填充每个节点的下一个右侧节点指针
 */

// @lc code=start
/**
 * Definition for a NextNode.
 * type NextNode struct {
 *     Val int
 *     Left *NextNode
 *     Right *NextNode
 *     Next *NextNode
 * }
 */

func connectNode1(left *NextNode, right *NextNode) {
	if left == nil || right == nil {
		return
	}

	left.Next = right
	connectNode1(left.Left, left.Right)
	connectNode1(right.Left, right.Right)
	connectNode1(left.Right, right.Left)
}

func connect1(root *NextNode) *NextNode {
	if root == nil {
		return root
	}
	connectNode1(root.Left, root.Right)
	return root
}

func Test116() {
	// 1,#,2,3,#,4,5,6,7,#
	fmt.Println(connect1(ToNode([]int{1, 2, 3, 4, 5, 6, 7})).ToValues())
}

// @lc code=end
